### PR TITLE
Added support for TElinear ops

### PR DESCRIPTION
**Pull request by kinjalpatel27**
iCo-authored-by: Asma Kuriparambil Thekkumpate <akuriparambi@nvidia.com>

## What does this PR do?

**Type of change:** New Feature

**Overview:** 
This MR adds support for quantizing TE Ops in megatron, specifically TERowParallelLinear, TEColumnParallelLinear and TELayerNormColumnParallelLinear.

## Usage
It can be used by enabling TE spec in megatron

## Testing
Added unit tests for testing functionality
`test_homogeneous_sharded_state_dict_te_spec`
`test_convert_mcore_te_gpt_model`
`test_quantize_forward_backward`

## Before your PR is "*Ready for review*"

- **Make sure you read and follow [Contributor guidelines](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CONTRIBUTING.md)** and your commits are signed.
- **Is this change backward compatible?**: Yes
- **Did you write any new necessary tests?**: Yes
- **Did you add or update any necessary documentation?**: Yes
- **Did you update [Changelog](https://github.com/NVIDIA/TensorRT-Model-Optimizer/blob/main/CHANGELOG.rst)?**: Yes

### CODE REVIEW

File: modelopt/torch/quantization/plugins/transformer_engine.py [hunk 1]
```diff
@@ -116,3 +150,118 @@ def te_grouped_quantized_linear_fn(package, func_name, self, *args):
...
+        if getattr(self, "fuse_wgrad_accumulation", False):
+            warnings.warn(
+                "fuse_wgrad_accumulation is not supported with ModelOpt quantization. "
+                "Setting fuse_wgrad_accumulation to False."
+            )
+            self.fuse_wgrad_accumulation = False
+
+    def forward(self, *args, **kwargs):
+        """Call ModelOpt patch for _LayerNormLinear functional."""
+        _QuantLayerNormLinearFunc.modelopt_quantizers = (
```
Comment (realAsma) [inline modelopt/torch/quantization/plugins/transformer_engine.py:272]: could you please add a comment that this is not multi-process safe, but not thread safe?

Comment (realAsma) [reply modelopt/torch/quantization/plugins/transformer_engine.py:272]: This works because at a time only one CPU process will be inside _QuantTELayerNormLinear.forward (i.e, it should not be a multi-thread process)

I previously saved the tensor_quantizers to input tensor - but that cause Torch Dynamo to break.

**Review by realAsma — APPROVED** [review]

File: modelopt/torch/quantization/plugins/transformer_engine.py (hunk 1 above)
Comment (kinjalpatel27) [reply modelopt/torch/quantization/plugins/transformer_engine.py:272]: done [afcbb21](https://github.com/NVIDIA/TensorRT-Model-Optimizer/pull/632/commits/afcbb21de6f837f51217ef6409c224c3010c2b57)

File: tests/gpu/torch/quantization/plugins/test_megatron.py [hunk 2]
```diff
@@ -932,3 +953,57 @@ def test_kv_cache_sharded_state_dict(tmp_path, config):
         job=partial(_test_kv_cache_sharded_state_dict_helper, tmp_path, config),
         backend="nccl",
     )
+
+
+def test_convert_mcore_te_gpt_model(distributed_setup_size_1):
```
Comment (mxinO) [inline tests/gpu/torch/quantization/plugins/test_megatron.py:965]: Do we have basic tests for TE? ref_output == test_output for quantizers disabled, and ref_output != test_output for quantizers enabled?

File: modelopt/torch/quantization/plugins/transformer_engine.py [hunk 3]
```diff
@@ -15,44 +15,71 @@
...
-                self.input_quantizer(inputs),
-                *remaining_args,
+                *new_args,
                 **kwargs,
             )
         else:
-            weight, weight_fp8, inputs = args[0], args[1], args[2]
-            remaining_args = args[3:]
+            idx = 1 if func_name == "_forward" else 0
+            weight, weight_fp8, inputs = args[idx], args[idx + 1], args[idx + 2]
```
Comment (mxinO) [inline modelopt/torch/quantization/plugins/transformer_engine.py:91]: We should raise an error somewhere if TE fp8 training is enabled when using modelopt.

**Review by mxinO — COMMENTED** [review]
Looks great to me! Do we still need a special modelopt layer spec after this MR?

File: modelopt/torch/quantization/plugins/transformer_engine.py (hunk 3 above)
Comment (kinjalpatel27) [reply modelopt/torch/quantization/plugins/transformer_engine.py:91]: fixed it in [d7c4802](https://github.com/NVIDIA/Model-Optimizer/pull/632/commits/d7c4802d3c69ae555ac6cf9917a22212adb836ee)

File: tests/gpu/torch/quantization/plugins/test_megatron.py (hunk 2 above)
Comment (kinjalpatel27) [reply tests/gpu/torch/quantization/plugins/test_megatron.py:965]: fixed it in [d7c4802](https://github.com/NVIDIA/Model-Optimizer/pull/632/commits/d7c4802d3c69ae555ac6cf9917a22212adb836ee)

File: modelopt/torch/quantization/plugins/transformer_engine.py [hunk 4]
```diff
@@ -15,44 +15,87 @@
...
+
+from modelopt.torch.quantization.utils import replace_function
 
 from ..nn import QuantModuleRegistry
 from .custom import _ParallelLinear
 
+_TE_VERSION = Version(te.__version__)
+
+
+def _check_te_fp8_enabled():
```
Comment (meenchen) [inline modelopt/torch/quantization/plugins/transformer_engine.py]: nit: maybe rename this as `_assert_te_fp8_disabled.`

File: modelopt/torch/quantization/plugins/transformer_engine.py [hunk 5]
```diff
@@ -116,3 +167,120 @@ def te_grouped_quantized_linear_fn(package, func_name, self, *args):
 
     # Override the quantized linear function
     _quantized_linear_fn = te_grouped_quantized_linear_fn
+
+
+class _QuantLayerNormLinearFunc(torch.autograd.Function):
```
Comment (meenchen) [inline modelopt/torch/quantization/plugins/transformer_engine.py:172]: qq, why do we need to quantize layernorm?

File: modelopt/torch/quantization/plugins/transformer_engine.py (hunk 4 above)
Comment (kinjalpatel27) [reply modelopt/torch/quantization/plugins/transformer_engine.py]: Updated it [a694984](https://github.com/NVIDIA/Model-Optimizer/pull/632/commits/a694984e3b1b49fe6c778c83924a65c9115a15bf)

File: modelopt/torch/quantization/plugins/transformer_engine.py (hunk 5 above)
Comment (kinjalpatel27) [reply modelopt/torch/quantization/plugins/transformer_engine.py:172]: We are quantizing [LayerNormLinear](https://docs.nvidia.com/deeplearning/transformer-engine/user-guide/api/pytorch.html#transformer_engine.pytorch.LayerNormLinear) function which Applies layer normalization followed by linear transformation to the incoming data, we are quantizing linear part.

Comment (meenchen) [reply modelopt/torch/quantization/plugins/transformer_engine.py:172]: I see.

**Comment by kinjalpatel27** [comment]
Thanks @meenchen 
> I am wondering if you have some data points of enabling TE in terms of perf and accuracy. 

@realAsma, do you have data points in terms of perf and accuracy for TE?

>Also, do we need to update the modelopt example in megatron to enable this by default?

We will enable TE by default in megatron and remove local spec to keep things consistent between megatron-core and megatron + modelopt example

**Comment by realAsma** [comment]
> I am wondering if you have some data points of enabling TE in terms of perf and accuracy.
@meenchen 

1. TELinear support will help us get rid of the ModelOpt model provider and simplify and clean up MCore flows - So we would be able to quantize any MCore model without having to explicitly convert to local spec. 
2. TELayerNorm Linear reduces memory consumption by orders of magnitude because of sequence parallelism.  With TELayerNorm we dont checkpoint the intermediate activations between layer norm and linear.
